core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,353 @@
-        new_unprocessed_indexes: &[usize],
-    ) -> bool {
-        let has_more_unprocessed_transactions = !new_unprocessed_indexes.is_empty();
-        if has_more_unprocessed_transactions {
-            self.unprocessed_packets
-                .retain(|index, _| new_unprocessed_indexes.contains(index));
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut DeserializedPacket> {
+        self.message_hash_to_transaction.iter_mut().map(|(_k, v)| v)
+    }
+
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut DeserializedPacket) -> bool,
+    {
+        // TODO: optimize this only when number of packets
+        // with oudated blockhash is high
+        self.packet_priority_queue.clear();
+        self.message_hash_to_transaction
+            .retain(|_k, deserialized_packet| {
+                let should_retain = f(deserialized_packet);
+                if should_retain {
+                    let priority_queue_entry =
+                        PacketPriorityQueueEntry::from_packet(deserialized_packet);
+                    self.packet_priority_queue.push(priority_queue_entry);
+                }
+                should_retain
+            })
+    }
+
+    pub fn len(&self) -> usize {
+        self.packet_priority_queue.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.packet_priority_queue.is_empty()
+    }
+
+    fn push_internal(&mut self, deserialized_packet: DeserializedPacket) {
+        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
+
+        // Push into the priority queue
+        self.packet_priority_queue.push(priority_queue_entry);
+
+        // Keep track of the original packet in the tracking hashmap
+        self.message_hash_to_transaction
+            .insert(deserialized_packet.message_hash(), deserialized_packet);
+    }
+
+    /// Returns the popped minimum packet from the priority queue.
+    fn push_pop_min(&mut self, deserialized_packet: DeserializedPacket) -> DeserializedPacket {
+        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
+
+        // Push into the priority queue
+        let popped_priority_queue_entry = self
+            .packet_priority_queue
+            .push_pop_min(priority_queue_entry);
+
+        // Keep track of the original packet in the tracking hashmap
+        self.message_hash_to_transaction
+            .insert(deserialized_packet.message_hash(), deserialized_packet);
+
+        // Remove the popped entry from the tracking hashmap. Unwrap call is safe
+        // because the priority queue and hashmap are kept consistent at all times.
+        self.message_hash_to_transaction
+            .remove(&popped_priority_queue_entry.message_hash)
+            .unwrap()
+    }
+
+    pub fn pop_max(&mut self) -> Option<DeserializedPacket> {
+        self.packet_priority_queue
+            .pop_max()
+            .map(|priority_queue_entry| {
```

Comment:
The unwrap is an assert that the entry that exists in `packet_priority_queue` must also exist in `message_hash_to_transaction`, whereas `and_then` would allow that to silently fail.

Reply:
We'll take the LoA bounty out of your salary :wink: